File: Python/NextGreaterElementI/NextGreaterElementI.py
# coding:utf-8
class Solution(object):
    # (62ms)
    def nextGreaterElement(self, findNums, nums):
        """
        :type findNums: List[int]
        :type nums: List[int]
        :rtype: List[int]
        """
        stack, num_map = [], {}
        for num in nums:
            while stack and stack[-1] < num:
                num_map[stack.pop()] = num
            stack.append(num)
        for index, item in enumerate(findNums):
            findNums[index] = num_map.get(item, -1)
        return findNums

    # (192ms)
    # 从python这两个运行效率来看，java中的map实现的是较差的
    def nextGreaterElement2(self, findNums, nums):
        ans = []
        for num in findNums:
            i = 0
            while nums[i] != num:
                i += 1
            i += 1
            while i < len(nums):
                if nums[i] > num:
                    ans.append(nums[i])
                    break
                i += 1
            if i == len(nums):
                ans.append(-1)
        return ans

print Solution().nextGreaterElement([2, 4], [1, 2, 3, 4])
print Solution().nextGreaterElement2([2, 4], [1, 2, 3, 4])
print Solution().nextGreaterElement([4, 1, 2], [1, 3, 4, 2])
print Solution().nextGreaterElement2([4, 1, 2], [1, 3, 4, 2])

File: JavaScript/NextGreaterElementI/NextGreaterElementI.js
/**
 * @param {number[]} findNums
 * @param {number[]} nums
 * @return {number[]}
 */
 // (192ms)
var nextGreaterElement = function(findNums, nums) {
    var greater = new Map();
	var stack = []
	nums.forEach(function(item){
		while(stack.length > 0 && stack[stack.length - 1] < item) {
			greater.set(stack.pop(), item);
		}
		stack.push(item);
	});
	findNums.forEach(function(item, index, array){
		if(greater.get(item)) {
			array[index] = greater.get(item);
		} else {
			array[index] = -1;
		}
	});
	return findNums;
};

// (142ms)
// 直接使用{}比使用Map效率更高
var nextGreaterElement = function(findNums, nums) {
    var greater = {};
	var stack = []
	nums.forEach(function(item){
		while(stack.length > 0 && stack[stack.length - 1] < item) {
			greater[stack.pop()] = item;
		}
		stack.push(item);
	});
	findNums.forEach(function(item, index, array){
		array[index] = greater[item] ? greater[item] : -1;
	});
	return findNums;
};

// (129ms)
// javascript的map实现的也很差
var nextGreaterElement2 = function(findNums, nums) {
    var ans = [];
    findNums.forEach(function(item){
        var i = 0;
        while(nums[i] != item) {
            i++;
        }
        for(i = i + 1; i < nums.length; i++) {
            if(nums[i] > item) {
                ans.push(nums[i]);
                break;
            }
        }
        if(i == nums.length) {
            ans.push(-1);
        }
    });
	return ans;
};